### PR TITLE
Commonize shim module join and shuffle files

### PR DESCRIPTION
**Pull request by tgravescs**
Many of the join and shuffle related files in the shim module were just there to build against the specific version of spark.  Moves those files into the sql-plugin.  Most of those had differences between 3.0.x and 3.1.x and then many times databricks had slight variations, so consolidated the shims module files into 3 or 4 in the sql-plugin directory.

Also note more to commonize later, doing it in pieces to hopefully keep easier to review.  

I ran tests on databricks and some Apache but not all.

### CODE REVIEW

**Comment by tgravescs** [comment]
build

**Comment by tgravescs** [comment]
databricks build failed because of https://github.com/NVIDIA/spark-rapids/pull/3556

**Comment by tgravescs** [comment]
build

**Review by gerashegalov — APPROVED** [review]
LGTM

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuBroadcastNestedLoopJoinExec.scala [hunk 1]
```diff
@@ -14,9 +14,10 @@
  * limitations under the License.
  */
 
-package com.nvidia.spark.rapids.shims.spark311
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuBroadcastNestedLoopJoinExec.scala:17]: Now that `GpuBroadcastNestedLoopJoinExec` is no longer shimmed in any way, should this be moved into the same file as `GpuBroadcastNestedLoopJoinExecBase`?  Is there still a reason for the latter class?

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastExchangeExec.scala [hunk 2]
```diff
@@ -356,7 +356,7 @@ abstract class GpuBroadcastExchangeExecBase(
         }
       }
     }
-    GpuBroadcastExchangeExec.executionContext.submit[Broadcast[Any]](task)
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastExchangeExec.scala:359]: Nit: Would like to see this file named `GpuBroadcastExchangeExecBase` or something more accurate to what it contains.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuShuffleExchangeExec.scala [hunk 3]
```diff
@@ -230,7 +230,7 @@ abstract class GpuShuffleExchangeExecBase(
     }
 }
 
-object GpuShuffleExchangeExec {
+object GpuShuffleExchangeExecBase {
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuShuffleExchangeExec.scala:233]: Nit: Update filename to GpuShuffleExchangeExecBase.scala.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuBroadcastNestedLoopJoinExec.scala (hunk 1 above)
Comment (tgravescs) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuBroadcastNestedLoopJoinExec.scala:17]: there is definitely more things to move around and commonize, I was trying to keep it simple here and easy to review by just moving the files out of shim modules without a lot of code changes that could add bugs and make it harder to review.   We can then follow up with more changes.

**Comment by tgravescs** [comment]
build

**Comment by tgravescs** [comment]
very odd > 07:22:08  [ERROR] /home/ubuntu/spark-rapids/tests/src/test/scala/com/nvidia/spark/rapids/SerializationSuite.scala:20: object lang3 is not a member of package org.apache.commons

builds fine locally and didn't change anything that would cause this

**Comment by tgravescs** [comment]
that failed on databricks and I bet is from https://github.com/NVIDIA/spark-rapids/pull/3504

**Comment by tgravescs** [comment]
build

**Comment by jlowe** [comment]
> rapids/tests/src/test/scala/com/nvidia/spark/rapids/SerializationSuite.scala:20: object lang3 is not a member of package org.apache.commons

Looks like the test was implicitly picking up commons lang3 from Spark and should have explicitly specified the dependency.  I'll post a PR to fix.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuBroadcastNestedLoopJoinExec.scala (hunk 1 above)
Comment (jlowe) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuBroadcastNestedLoopJoinExec.scala:17]: I'm fine with followups as long as we're tracking somewhere the things we want to do so they don't get dropped.

File: pom.xml [hunk 4]
```diff
@@ -263,7 +267,9 @@
                                         <source>${project.basedir}/src/main/301until320-all/scala</source>
                                         <source>${project.basedir}/src/main/311until320-all/scala</source>
                                         <source>${project.basedir}/src/main/311+-all/scala</source>
+                                        <source>${project.basedir}/src/main/311+-nondb/scala</source>
                                         <source>${project.basedir}/src/main/311until320-apache/scala</source>
+                                        <source>${project.basedir}/src/main/301+-nondb/scala</source>
```
[the remark this answers is not in the document]
Comment (tgravescs) [reply pom.xml:272]: also note I have older ide which doesn't have that option.  I'll look at upgrading later, but I'm not sure I agree with its sorting going by the other ones, for instance:

```
<source>${project.basedir}/src/main/301until310-nondb/scala</source>
<source>${project.basedir}/src/main/301until320-nondb/scala</source>
<source>${project.basedir}/src/main/301until320-all/scala</source>
 <source>${project.basedir}/src/main/301until310-all/scala</source>
 <source>${project.basedir}/src/main/301+-nondb/scala</source>
```

Why is it putting nondb before all, doesn't seem like by alphabetical? As long as we are consistent I'm fine with it but don't want to require certain version of IDE either though.  Or perhaps those weren't sorted?

**Review by jlowe — APPROVED** [review]

**Pull request merged**